File: leetcode/401-500/405-convert-a-number-to-hexadecimal/bench/to_hex.py
# LeetCode #405 bench harness — Python peer (CPython, single-thread).
#
# Bitwise mask-and-shift hex render — same canonical algorithm as the Kara
# mirror. Sequential string-building bench: concatenate TOTAL hex renderings
# into one buffer, then byte-checksum it. Slowest lane; bounds the interpreter
# cost. Sink = byte-sum of the concatenated output.

TOTAL = 4000000
HEX = "0123456789abcdef"


def to_hex(num):
    n = num & 0xFFFFFFFF
    if n == 0:
        return "0"
    digits = []
    while n > 0:
        digits.append(HEX[n & 0xF])
        n >>= 4
    return "".join(reversed(digits))


def main():
    parts = []
    for k in range(TOTAL):
        v = (k * 2654435761) & 0xFFFFFFFF
        parts.append(to_hex(v))
    out = "".join(parts)
    print(sum(ord(c) for c in out))


main()

File: leetcode/401-500/405-convert-a-number-to-hexadecimal/bench/to_hex.rs
// LeetCode #405 bench harness — Rust peer (rustc -O, single-thread).
//
// Bitwise mask-and-shift hex render — same canonical algorithm as the Kāra
// mirror. Sequential string-building bench: concatenate TOTAL hex renderings
// into one buffer, then byte-checksum it. Persisting the output defeats
// allocation-elision (a per-render byte-sum would let rustc fold the work to
// arithmetic). Sink = byte-sum of the concatenated output.

const TOTAL: i64 = 4_000_000;
const HEX: &[u8; 16] = b"0123456789abcdef";

fn to_hex(num: i64) -> String {
    if num == 0 {
        return "0".to_string();
    }
    let mut n = num & 0xffff_ffff;
    let mut buf: Vec<i64> = Vec::new();
    while n > 0 {
        buf.push(n & 0xf);
        n >>= 4;
    }
    let mut out = String::new();
    for k in (0..buf.len()).rev() {
        out.push(HEX[buf[k] as usize] as char);
    }
    out
}

fn main() {
    let mut out = String::new();
    for k in 0..TOTAL {
        let v = (k * 2654435761) & 0xffff_ffff;
        out.push_str(&to_hex(v));
    }
    let mut sum: i64 = 0;
    for &b in out.as_bytes() {
        sum += b as i64;
    }
    println!("{}", sum);
}
